animation-proposal.md

```diff
@@ -0,0 +1,338 @@
+	| "relative"
+	;
+
+/**
+ * Describes how many times the interpolation should repeat.
+ */
+export type InterpolationIterationCount
+	/** Run the interpolation this many times. */
+	= number
+
+	/** Run the interpolation forever (until canceled). */
+	| "infinite"
+	;
+
+/**
+ * Specifies the direction the interpolation should be played.
+ */
+export type InterpolationDirection
+	/** The interpolation is played forward. */
+	= "forward"
+	/** The interpolation is played in reverse. */
+	| "reverse"
+	/** The interpolation is played forward, then backward, alternating. */
+	| "alternate"
+	/** Same as "alternate", but starts out playing backward. */
+	| "alternate-reverse"
+	;
+```
+
+##### Examples
+
+```ts
+// Interpolate a vector (lerp).
+this.context.interpolate(actor.selector("transform.app.position"), position, 0.2);
+
+// Interpolate a quaternion (slerp).
+this.context.interpolate(actor.selector("transform.app.rotation"), rotation, 0.2, { curve: "ease" });
+
+// Interpolate a number.
+this.context.interpolate(actor.selector("light.intensity"), 1, 0.2, { curve: "ease-in" });
+
+// Interpolate a color.
+this.context.interpolate(material.selector("color"), color, 0.2, { curve: "ease-out" });
+
+// Interpolate a value on a custom curve (a "bounce", in this case).
+this.context.interpolate(material.selector("color"), color, 0.2, { curve: [0.1, -0.6, 0.2, 0] });
+
+// Interpolate to another actor's position by reference. The host reads the
+// target actor's position locally via the target selector, enabling smooth
+// continuous interpolation to a moving target, and alleviating the need to
+// subscribe to the target actor's transform.
+this.context.interpolate(
+	actor.selector("transform.app.position"),
+	target.selector("transform.app.position"),
+	0.2, {
+		iterationCount: "infinite"
+	});
+
+// Cancel an interpolation. This applies to all types of interpolations, but
+// is particularly useful for stopping continuous interpolations.
+this.context.interpolate(
+	actor.selector("transform.app.position"),
+	target.selector("transform.app.position"),
+	null);
+)
+```
+
+#### Add `Actor.selector` and `Asset.selector` methods
+
+Add a `selector` method to classes `Actor` and `Asset`. `selector` takes a string representation of a path to a member and returns an absolute path to it, allowing the host to look up the value client-side.
+
+```ts
+/**
+ * Returns an absolute reference to the given field.
+ * @param path The relative path to the field, e.g.: "transform.local.position".
+ * @returns An absolute path to the field, suitable for use by the host to look
+ * up the value client-side.
+ */
+public selector(path: string): string;
+```
+
+
+#### Convenience API
+
+An `interpolate` method on context isn't very discoverable. We may want to add convenience methods on `Actor` and `Asset`, so that `interpolate` shows up in intellisense. Example:
+
```

Comment:
Why have this as a "convenience API", and not just leave this as the only option? Why would anyone want to use context.interpolate(actor.selector())?. That may be how it works internally, but there's no benefit for end users to worry about that, right?

Reply:
Yep that would be clearer/simpler.

Reply:
Turns out there is an issue with moving `interpolate` off `Context`. Reverting that change for now and let's chat in person.